text/metrics/0000-pipeline-monitoring.md

```diff
@@ -0,0 +1,347 @@
+
+Users are expected to understand that the data item for traces is a
+span, for logs is a record, and for metrics is a point.
+
+#### Pipeline component name
+
+Components are uniquely identified using a descriptive `name`
+attribute which encompasses at least a short name describing the type
+of component being used (e.g., `batch` for the SDK BatchSpanProcessor
+or the Collector batch proessor).
+
+When there is more than one component of a given type active in a
+pipeline having the same `domain` and `signal` attributes, the `name`
+should include additional information to disambiguate the multiple
+instances using the syntax `<type>/<instance>`.  For example, if there
+were two `batch` processors in a collection pipeline (e.g., one for
+error spans and one for non-error spans) they might use the names
+`batch/error` and `batch/noerror`.
+
+#### Pipeline monitoring diagram
+
+The relationship between items received, dropped, and exported is
+shown in the following diagram.
+
+![pipeline monitoring metrics](../images/otel-pipeline-monitoring.png)
+
+### Proposed metrics semantic conventions
+
+The proposed metric names are: 
+
+`otel.{station}.received`: Inclusive count of items entering the pipeline at a station.
+`otel.{station}.dropped`: Non-inclusive count of items dropped by a component in a pipeline.
+`otel.{station}.exported`: Inclusive count of items exiting the pipeline at a station.
+
+The behavior specified for SDKs and Collectors at each level of detail
+is different, because SDKs do not receive items from a pipeline.
+
+#### SDK default configuration
+
+At the basic level of detail, SDKs are required to count spans
+received by the export pipeline.  Only the `otel.sdk.received` metric
+is required.  This includes items that succeed, fail, or are dropped.
+
+At the normal level of detail, the `otel.sdk.received` metric gains an
+additional boolean attribute, `success` indicating success or failure.
+Also at the normal level of detail, the `otel.sdk.dropped` metric is
+counted.
+
+#### Collector default configuration
+
+At the basic level of detail, Collectors of a given `{station}` name
+are required to count `otel.{station}.received`,
+`otel.{station}.dropped`, and `otel.{station}.exported`.
+
+At the normal level of detailed, the `received` and `exported` metrics
+gain an additional boolean attribute, `success` indicating success or
+failure.
+
+#### Detailed-level metrics configuration
+
+There is one additional dimension that users may wish to opt-in to, in
+order to gain information about failures at a particular pipeline
+stage.  When detail-level metrics are requested, all three metric
+instruments specified for pipeline monitoring gain an additional
+`reason` attribute, with a short string explaining the failure.  
+
+For example, with detailed-level metrics in use, the
+`otel.{station}.received` and `otel.{station}.exported` counters will
+include additional `reason` information (e.g., `timeout`,
```

Comment:
I've been thinking about generally how domain-specific export metrics should relate to protocol-specific metrics.  otelhttp and otelgrpc should, in theory, be the canonical way to get a reason for receive/export failures, but that means looking at multiple metrics, and knowing which protocol the exporter uses. Would it be better for otelhttp/otelgrpc telemetry to be the "detailed" mechanism for finding the reason for failures?

Reply:
This is a top-of-mind question for me. I would definitely sacrifice detailed-level metrics from the pipeline-monitoring logic in favor of transport-specific metrics, especially when a user is not mixing transports. I am not sure this argument works quite as well for collectors.

I worry that the details are quite a bit different too. When I use gRPC-specific instrumentation, I can't be sure that the pipeline-monitoring equations stated in this document will work (e.g., how are retries counted? do those specifications exist yet?)--they could help diagnose different sorts of problem, but the metrics specific to a transport will not easily integrate with the pipeline-monitoring logic outlined here.